autogpt/commands/file_operations.py

```diff
@@ -17,30 +18,92 @@
 CFG = Config()
 
 
-def check_duplicate_operation(operation: str, filename: str) -> bool:
-    """Check if the operation has already been performed on the given file
+def text_checksum(text: str) -> str:
+    """Get the hex checksum for the given text."""
+    digest = hashlib.md5()
+    digest.update(text.encode("utf-8"))
+    return digest.hexdigest()
+
+
+def operations_from_log(log_path: str) -> Generator[Tuple]:
+    with open(log_path, "r", encoding="utf-8") as log:
+        for line in log:
+            line = line.replace("File Operation Logger", "").strip()
+            if not line:
+                continue
+            operation, tail = line.split(": ", maxsplit=1)
+            operation = operation.strip()
+            if operation in ("write", "append"):
+                try:
+                    path, checksum = (x.strip() for x in tail.rsplit(" #", maxsplit=1))
+                except ValueError:
+                    path, checksum = tail.strip(), None
+                yield (operation, path, checksum)
+            elif operation == "delete":
+                yield (operation, tail.strip(), None)
+
+
+def file_operations_state(log_path: str) -> Dict:
+    """Iterates over the operations log and returns the expected state.
+
+    Parses a log file at CFG.file_logger_path to construct a dictionary that maps
+    each file path written or appended to its checksum. Deleted files are removed
+    from the dictionary.
+
+    Returns:
+        A dictionary mapping file paths to their checksums.
+
+    Raises:
+        FileNotFoundError: If CFG.file_logger_path is not found.
+        ValueError: If the log file content is not in the expected format.
+    """
+    state = {}
+    for operation, path, checksum in operations_from_log(log_path):
+        if operation in ("write", "append"):
+            state[path] = checksum
+        elif operation == "delete":
+            del state[path]
+    return state
+
+
+def new_entry(operation: str, filename: str, checksum: str | None = None) -> str:
+    """Returns the formatted log entry"""
+    if checksum is None:
+        return f"{operation}: {filename}\n"
+    return f"{operation}: {filename} #{checksum}\n"
+
+
+def is_duplicate_operation(
```

Comment:
This could be a decorator `@prevent_duplicates`:
* wraps the decorated function
* checking for duplicate operations before executing
* and logging it afterward

Would look a lot cleaner in the code :)

Reply:
i agree, but i think it'd be better as a separate PR because this is pretty significant as-is. plus the behavior of writes and deletes are _just_ different enough that I couldn't just slap it on